### PR TITLE
add toBuilder method on some models

### PR DESCRIPTION
**Pull request by channel-dante**
### Description
- added toBuilder method to all request model in core package & _types.query_dsl package
- for consistent model management, add to all related models rather than just a specific model.

### Issues Resolved
- https://github.com/opensearch-project/opensearch-java/issues/583

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
This looks good. Any way we can get some test coverage here on these other classes. I understand it's laborious (hence we should be investing in a code generator).

**Comment by channel-dante** [comment]
> This looks good. Any way we can get some test coverage here on these other classes? I understand it's laborious (hence we should be investing in a code generator).


Would a simple test like [this](https://github.com/opensearch-project/opensearch-java/blob/344d8ae2b88594afcf85a16274f64b290468e615/java-client/src/test/java/org/opensearch/client/opensearch/core/SearchRequestTest.java) be enough? Or is it necessary to test on all fields?

**Comment by dblock** [comment]
> > This looks good. Any way we can get some test coverage here on these other classes? I understand it's laborious (hence we should be investing in a code generator).
> 
> @dblock I'm curious what you want to test. Can I just test whether toBuilder works ([like this](https://github.com/opensearch-project/opensearch-java/blob/344d8ae2b88594afcf85a16274f64b290468e615/java-client/src/test/java/org/opensearch/client/opensearch/core/SearchRequestTest.java))? Or do you want to verify that all fields operate normally? If you want a second one, it won't be an easy task.

I'm good with anything that just exercises the code to start.

**Comment by channel-dante** [comment]
@dblock Can you please review again? It was a very laborious task 😅

File: CHANGELOG.md [hunk 1]
```diff
@@ -39,12 +39,15 @@ This section is for maintaining a changelog for all breaking changes for the cli
 - Added support for icu_collation_keyword type ([#725](https://github.com/opensearch-project/opensearch-java/pull/725))
 - Added support for flat_object field property ([#735](https://github.com/opensearch-project/opensearch-java/pull/735))
 - Added toQuery method in Query and QueryVariant ([#760](https://github.com/opensearch-project/opensearch-java/pull/760)
+- Added toBuilder method to all request model in core package & _types.query_dsl package
```
Comment (dblock) [inline CHANGELOG.md:42]: Add a link to the PR like other lines.

**Review by dblock — APPROVED** [review]
Looks good to me! Nit on CHANGELOG.

Thank you for doing this. I recommend investing more time in code generation so we don't have to do this by hand :)

**Comment by dblock** [comment]
@VachaShah @reta Any ideas of how to do this kind of change better?

**Comment by VachaShah** [comment]
> @VachaShah @reta Any ideas of how to do this kind of change better?

I think for changes like these which span almost all files, code generation is the way to go. If we have some code which can incrementally add such methods (even if it is not a complete code generator for the client), it would be beneficial to add features like `toString` and `withJson`.

**Review by VachaShah — DISMISSED** [review]
This is great! Thank you @channel-dante for adding this along with the tests!

**Comment by reta** [comment]
> @VachaShah @reta Any ideas of how to do this kind of change better?

Agree with @VachaShah , having code generation in place would make it much simpler ....

**Review by dblock — DISMISSED** [review]

**Comment by channel-dante** [comment]
@VachaShah Can you please confirm about this pr?

**Comment by dblock** [comment]
@channel-dante rebase?

**Review by VachaShah — APPROVED** [review]
LGTM, thank you @channel-dante!

**Comment by dblock** [comment]
@channel-dante you should backport this manually to 2.x so we can release it with the next minor

**Pull request merged**